File: src/utils/validators.js
/**
 * Request Validators
 * Joi schemas for validating incoming requests.
 */
const Joi = require('joi');

// ---- /upload validation ----
const uploadSchema = Joi.object({
  // File is validated via Multer; this covers optional body fields
  chunkSize: Joi.number().integer().min(100).max(10000).optional(),
  chunkOverlap: Joi.number().integer().min(0).max(500).optional(),
  metadata: Joi.object().optional()
});

// ---- /query validation ----
const querySchema = Joi.object({
  query: Joi.string().trim().min(1).max(5000).required()
    .messages({ 'string.empty': 'Query text is required' }),
  topK: Joi.number().integer().min(1).max(20).default(5),
  documentId: Joi.string().uuid().optional(),
  includeMetadata: Joi.boolean().default(true),
  temperature: Joi.number().min(0).max(1).optional()
});

// ---- /compare validation ----
const compareSchema = Joi.object({
  documentIds: Joi.array()
    .items(Joi.string().uuid())
    .length(2)
    .required()
    .messages({
      'array.length': 'Exactly two document IDs are required for comparison'
    }),
  topic: Joi.string().trim().min(1).max(2000).required()
    .messages({ 'string.empty': 'Comparison topic is required' }),
  topK: Joi.number().integer().min(1).max(20).default(5)
});

// ---- /documents/:id param validation ----
const documentIdSchema = Joi.object({
  id: Joi.string().uuid().required()
});

/**
 * Express middleware factory: validates req.body against the given schema.
 * @param {Joi.ObjectSchema} schema
 * @returns {Function} Express middleware
 */
function validate(schema) {
  return (req, res, next) => {
    const { error, value } = schema.validate(req.body, {
      abortEarly: false,
      stripUnknown: true,
      convert: true
    });

    if (error) {
      const details = error.details.map(d => ({
        field: d.path.join('.'),
        message: d.message
      }));
      return res.status(400).json({
        success: false,
        error: {
          message: 'Validation failed',
          details
        }
      });
    }

    // Replace body with validated & converted values
    req.body = value;
    next();
  };
}

/**
 * Validate request params.
 */
function validateParams(schema) {
  return (req, res, next) => {
    const { error, value } = schema.validate(req.params, {
      abortEarly: false,
      stripUnknown: true
    });

    if (error) {
      const details = error.details.map(d => ({
        field: d.path.join('.'),
        message: d.message
      }));
      return res.status(400).json({
        success: false,
        error: {
          message: 'Validation failed',
          details
        }
      });
    }

    req.params = value;
    next();
  };
}

module.exports = {
  uploadSchema,
  querySchema,
  compareSchema,
  documentIdSchema,
  validate,
  validateParams
};

File: app/utils/validators.py
"""
Request Validators
Pydantic models for validating incoming requests.
"""
from typing import List, Optional
from pydantic import BaseModel, Field, field_validator
import uuid


class UploadParams(BaseModel):
    """Optional body fields sent alongside file upload."""
    chunk_size: Optional[int] = Field(None, ge=100, le=10000)
    chunk_overlap: Optional[int] = Field(None, ge=0, le=500)


class QueryRequest(BaseModel):
    """POST /api/query body."""
    query: str = Field(..., min_length=1, max_length=5000)
    top_k: int = Field(5, ge=1, le=20)
    document_id: Optional[str] = None
    include_metadata: bool = True
    temperature: Optional[float] = Field(None, ge=0.0, le=1.0)
    verify: bool = False
    rerank: bool = False

    @field_validator("document_id")
    @classmethod
    def validate_uuid(cls, v):
        if v is None:
            return None
        # Treat Swagger placeholders and empty strings as "not provided"
        if not v or not v.strip() or v.strip().lower() == "string":
            return None
        try:
            uuid.UUID(v)
        except ValueError:
            raise ValueError("document_id must be a valid UUID")
        return v


class CompareRequest(BaseModel):
    """POST /api/compare body."""
    document_ids: List[str] = Field(..., min_length=2, max_length=2)
    topic: str = Field(..., min_length=1, max_length=2000)
    top_k: int = Field(5, ge=1, le=20)
    structured: bool = False

    @field_validator("document_ids")
    @classmethod
    def validate_doc_ids(cls, v):
        if len(v) != 2:
            raise ValueError("Exactly two document IDs are required for comparison")
        for doc_id in v:
            try:
                uuid.UUID(doc_id)
            except ValueError:
                raise ValueError(f"Invalid UUID: {doc_id}")
        return v
